pennylane/labs/trotter_error/product_formulas/error.py

```diff
@@ -63,44 +63,57 @@ def trotter_error(fragments: Sequence[Fragment], delta: float) -> Fragment:
     >>>     r_state.random(size=(n_modes, n_modes)),
     >>>     r_state.random(size=(n_modes, n_modes, n_modes))
     >>> ]
-    >>> frags = vibrational_fragments(n_modes, freqs, taylor_coeffs)
+    >>>
     >>> delta = 0.001
-    >>> type(trotter_error(frags, delta))
-    <class 'pennylane.labs.trotter_error.realspace.realspace_operator.RealspaceSum'>
+    >>> frag_labels = [0, 1, 1, 0]
+    >>> frag_coeffs = [1/2, 1/2, 1/2, 1/2]
 
+    >>> pf = ProductFormula(frag_labels, coeffs=frag_coeffs)
+    >>> frags = dict(enumerate(vibrational_fragments(n_modes, freqs, taylor_coeffs)))
+    >>> type(effective_hamiltonian(pf, frags, order=5, timestep=delta))
+    <class 'pennylane.labs.trotter_error.realspace.realspace_operator.RealspaceSum'>
     """
 
-    if len(fragments) == 0:
-        return fragments
+    if not product_formula.fragments.issubset(fragments.keys()):
+        raise ValueError("Fragments do not match product formula")
 
+    bch = bch_expansion(product_formula(1j * timestep), order)
     eff = _AdditiveIdentity()
-    n_frags = len(fragments)
-    scalar = -(delta**2) / 24
 
-    for i in range(n_frags - 1):
-        for j in range(i + 1, n_frags):
-            eff += nested_commutator([fragments[i], fragments[i], fragments[j]])
-            for k in range(i + 1, n_frags):
-                eff += 2 * nested_commutator([fragments[k], fragments[i], fragments[j]])
+    for ith_order in bch:
+        for commutator, coeff in ith_order.items():
+            eff += coeff * nested_commutator(_insert_fragments(commutator, fragments))
 
-    eff = scalar * eff
     return eff
 
 
+def _insert_fragments(commutator, fragments):
```

Comment:
Please add a description. The name suggests that it inserts a fragment to a commutator, but not sure the function does that.